### DISCUSSION TITLE
How to determine how langversion:latest is interpreted?

### DISCUSSION THREAD

**Discussion by mrlacey** [dotnet/roslyn#62093] [Q&A] [answered]
I'm working on a large solution with lots of complicated dependencies.
Everything works fine when we set `<LangVersion>latest</LangVersion>` or `<LangVersion>default</LangVersion>` explicitly in the project files. However, I get errors when I explicitly set `<LangVersion>10.0</LangVersion>`.

I'm using VS 17.2.3, but according to the compiler, this should be the same and not produce errors.

```ascii
C:\Program Files\Microsoft Visual Studio\2022\Community>csc -langversion:?
Supported language versions:
default
1
2
3
4
5
6
7.0
7.1
7.2
7.3
8.0
9.0
10.0 (default)
latestmajor
preview
latest
```

Is there a way to find out what the compiler is actually using when the language version is set as something other than an explicit number?

Given the order of versions listed above, does "latest" include everything in  "preview"? or is it deemed equivalent to "default"?

Given I'm using VisualStudio, is there a way to find out the supported (and used) language versions? (as with `csc.exe` above)

Additionally, I get different errors (some things work and different things don't) when I specify the "LangVersion" as either `10` or `10.0`.
Should these be treated the same (regardless of specifying the minor version number)?

The solution is also using a lot of SourceGenerators that are connected with the errors. Are there any known issues with the specified language version of a project affecting the way a source generator behaves? Or the language version the source generator uses?

Pre-empting the obvious question: This is a large, complex solution that I can't share publicly yet and I'm struggling to create a minimal repro. I will share as soon as I have something.

_I know this is a lot of questions (sorry) but I feel I'm pushing at the limits of what's documented._

**Comment by svick** [comment, marked as answer]
> Is there a way to find out what the compiler is actually using when the language version is set as something other than an explicit number?

If you add `#error version` to some C# file that you're compiling, the compiler will produce a diagnostic error that includes information about the language version used, e.g.:

> error CS1029: #error: 'version'
> error CS8304: Compiler version: '4.3.0-2.22307.7 (069a85a7)'. Language version: default (10.0).

---

> Given the order of versions listed above, does "latest" include everything in "preview"? or is it deemed equivalent to "default"?

[The documentation says:](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version#c-language-version-reference)

> Value | Meaning
> -- | --
> `preview` | The compiler accepts all valid language syntax from the latest preview version.
> `latest` | The compiler accepts syntax from the latest released version of the compiler (including minor version).
> `latestMajor` (`default`) | The compiler accepts syntax from the latest released major version of the compiler.

Since C# stopped using minor versions, `default` and `latest` are effectively the same and are different from `preview`.

---

> Given I'm using VisualStudio, is there a way to find out the supported (and used) language versions? (as with `csc.exe` above)

Most of the time, you shouldn't even need to set `LangVersion`, so VS doesn't make this easy.

Though there is an obscure and complicated way to do this using `<LangVersion>?</LangVersion>`, see https://github.com/dotnet/docs/issues/27101 for some details on how to actually see the output.

---

> Additionally, I get different errors (some things work and different things don't) when I specify the "LangVersion" as either `10` or `10.0`.
Should these be treated the same (regardless of specifying the minor version number)?

Yes, 10 and 10.0 are the same version. If you can reliably reproduce a difference between the two, consider creating an issue in this repo with the details.

**Comment by PathogenDavid** [comment]
> Is there a way to find out what the compiler is actually using when the language version is set as something other than an explicit number?

If you type `#error version` in an affected file, it print and error with the compiler and language versions on build. For example, here's what it gives with `<LangVersion>latest</LangVersion>` on my machine:

```
error CS8304: Compiler version: '4.3.0-2.22307.7 (069a85a7)'. Language version: latest (10.0).
```

(While not common, depending on what else is going on in the project, Visual Studio and MSBuild can have different versions of Roslyn so don't rely on the error you get when hovering it.)

> Given the order of versions listed above, does "latest" include everything in "preview"? or is it deemed equivalent to "default"?

`latest` does not include `preview`. [The documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/language) documents `default` as being equivalent to `latestMajor`. Since C# doesn't use minor versions anymore, `latest` and `default` mean the same thing.

You can see this reflected in the mapping of the non-numeric versions in [`MapSpecifiedToEffectiveVersion`](https://github.com/dotnet/roslyn/blob/69bdfb29ce4d027c5f5bf70f50be9bebcb5df1aa/src/Compilers/CSharp/Portable/LanguageVersion.cs#L465).

It is worth noting that `default` is the default from the perspective of Roslyn, not the SDK. IE: It doesn't know to prefer C# 9 for .NET 5 when C# 10 is available.

----------------

> *I know this is a lot of questions (sorry) but I feel I'm pushing at the limits of what's documented.*

You may find it helpful to use the [MSBuild Structured Log Viewer](https://github.com/KirillOsenkov/MSBuildStructuredLog) to inspect what's actually happening during your build. It will tell you (among many things) what specific copy of the compiler is being used by your project. Search for `$Task csc`.

If you suspect things are differing between command line builds and F7 builds in Visual Studio, you can use [the Project System Tools extension](https://github.com/dotnet/project-system-tools) to get a binlog from Visual Studio directly.

As an aside, make sure there's no uses of [`Microsoft.Net.Compilers.Toolset`](https://github.com/dotnet/roslyn/blob/69bdfb29ce4d027c5f5bf70f50be9bebcb5df1aa/docs/compilers/Compiler%20Toolset%20NuPkgs.md) in your project. It could cause an issue like this.

**Comment by mrlacey** [comment]
For reference, the issue I originally encountered turned out be caused by a package referenced in the project and was not an issue in the compiler. :(/:)
